File: packages/sdk-python/veltrix_sdk/resources/log_forwarding.py
# -*- coding: utf-8 -*-

from .base import BaseResource


class LogForwardingResource(BaseResource):
    """Handles log-forwarding destinations (``/api/log-forwarding``)."""
    RESOURCE_PATH = "log-forwarding"  # Note the hyphen

    def list(self, **kwargs):
        """Lists all log-forwarding destinations (GET /api/log-forwarding)."""
        return self._list(**kwargs)

    def create(self, name, forwarding_type, endpoint, **kwargs):
        """Creates a log-forwarding destination (POST /api/log-forwarding).

        Args:
            name: Display name of the destination.
            forwarding_type: Destination type (e.g. ``splunk``, ``datadog``, ``custom``).
            endpoint: The destination endpoint URL.
        """
        data = {"name": name, "type": forwarding_type, "endpoint": endpoint}
        return self._create(data=data, **kwargs)

    def update(self, destination_id, name=None, forwarding_type=None, endpoint=None,
               status=None, **kwargs):
        """Updates a log-forwarding destination (PUT /api/log-forwarding/{id})."""
        data = {
            "name": name,
            "type": forwarding_type,
            "endpoint": endpoint,
            "status": status,
        }
        data = {k: v for k, v in data.items() if v is not None}
        return self._update(resource_id=destination_id, data=data, **kwargs)

    def delete(self, destination_id, **kwargs):
        """Deletes a log-forwarding destination (DELETE /api/log-forwarding/{id})."""
        return self._delete(resource_id=destination_id, **kwargs)

    def test(self, destination_id, **kwargs):
        """Tests a log-forwarding destination (POST /api/log-forwarding/{id}/test)."""
        return self._action(resource_id=destination_id, action="test", method="POST", **kwargs)

File: packages/sdk-js/src/resources/log_forwarding.ts
import { BaseResource } from './base';
import { AxiosRequestConfig } from 'axios';
import { APIError } from '../errors';

// Define interfaces for Log Forwarding data structures
type LogForwardingType = 'splunk' | 'elasticsearch' | 'datadog' | 'sumologic' | 'custom';
type LogForwardingStatus = 'active' | 'inactive' | 'error';

interface LogForwardingDestination {
  id: string; // uuid
  name: string;
  type: LogForwardingType;
  endpoint: string;
  status: LogForwardingStatus;
  error?: string | null;
  lastSync?: string | null; // ISO Date string
  customerId: string; // uuid
  createdAt: string; // ISO Date string
  updatedAt: string; // ISO Date string
}

interface CreateLogForwardingPayload {
  name: string;
  type: LogForwardingType;
  endpoint: string;
}

interface UpdateLogForwardingPayload {
  name?: string;
  type?: LogForwardingType;
  endpoint?: string;
  status?: LogForwardingStatus;
}

interface TestLogForwardingResponse {
    success: boolean;
    message: string;
}

export class LogForwardingResource extends BaseResource {
  protected readonly RESOURCE_PATH = "log-forwarding"; // Note hyphen

  async list(config?: AxiosRequestConfig): Promise<LogForwardingDestination[]> {
    // Corresponds to GET /api/log-forwarding
    return this._list<LogForwardingDestination[]>(undefined, config);
  }

  async create(payload: CreateLogForwardingPayload, config?: AxiosRequestConfig): Promise<LogForwardingDestination> {
    // Corresponds to POST /api/log-forwarding
     const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this._create<LogForwardingDestination>(cleanedPayload, undefined, config);
  }

  // GET /api/log-forwarding/{id} is missing from spec, skipping get() method

  async update(destinationId: string, payload: UpdateLogForwardingPayload, config?: AxiosRequestConfig): Promise<LogForwardingDestination> {
    // Corresponds to PUT /api/log-forwarding/{id}
     const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this._update<LogForwardingDestination>(destinationId, cleanedPayload, undefined, config);
  }

  async delete(destinationId: string, config?: AxiosRequestConfig): Promise<{ message: string }> {
    // Corresponds to DELETE /api/log-forwarding/{id}
    // API Spec indicates 200 OK with message for delete
     const result = await this._delete<{ message: string } | null>(destinationId, undefined, config);
     if (result === null) {
        throw new APIError(`DELETE ${this._getPath(destinationId)} returned null/204, expected 200 with message.`);
    }
    return result;
  }

  async test(destinationId: string, config?: AxiosRequestConfig): Promise<TestLogForwardingResponse> {
    // Corresponds to POST /api/log-forwarding/{id}/test
    const action = "test";
    // Use _action helper as it's a POST to a sub-path with no body
    return this._action<TestLogForwardingResponse>(destinationId, action, 'POST', undefined, undefined, config);
  }
}
